File: backend/app/models/person.py
"""Modelos `Person` e `PersonDraft` — cadastro central de pessoas de interesse.

`Person` é fonte de verdade. Futuras features (1:1s, transcrições, feedbacks)
referenciam pela FK em `id`. `PersonDraft` é um buffer transient para propostas
do agente enriquecedor; promove para `Person` após confirmação do EM.
"""

from __future__ import annotations

import enum
import uuid
from datetime import date, datetime

from sqlalchemy import Date, DateTime, Enum, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.models import Base


class RelationshipType(str, enum.Enum):
    direct_report = "direct_report"
    peer = "peer"
    manager = "manager"
    skip_level = "skip_level"
    stakeholder = "stakeholder"
    other = "other"


def _uuid() -> str:
    return str(uuid.uuid4())


class Person(Base):
    __tablename__ = "people"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    email: Mapped[str] = mapped_column(String(200), nullable=False, unique=True, index=True)
    role: Mapped[str | None] = mapped_column(String(120), nullable=True)
    relationship: Mapped[RelationshipType] = mapped_column(
        Enum(RelationshipType, name="relationship_type"), nullable=False
    )
    slack_id: Mapped[str | None] = mapped_column(String(40), nullable=True, index=True)
    jira_account_id: Mapped[str | None] = mapped_column(String(80), nullable=True)
    github_handle: Mapped[str | None] = mapped_column(String(80), nullable=True)
    start_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )


class PersonDraft(Base):
    __tablename__ = "person_drafts"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    email: Mapped[str] = mapped_column(String(200), nullable=False, index=True)
    role: Mapped[str | None] = mapped_column(String(120), nullable=True)
    relationship: Mapped[RelationshipType] = mapped_column(
        Enum(RelationshipType, name="relationship_type"), nullable=False
    )
    slack_id: Mapped[str | None] = mapped_column(String(40), nullable=True)
    jira_account_id: Mapped[str | None] = mapped_column(String(80), nullable=True)
    github_handle: Mapped[str | None] = mapped_column(String(80), nullable=True)
    start_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    source: Mapped[str] = mapped_column(String(40), nullable=False, default="manual")
    proposed_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

File: frontend/src/types/person.ts
export type RelationshipType =
  | 'direct_report'
  | 'peer'
  | 'manager'
  | 'skip_level'
  | 'stakeholder'
  | 'other';

export interface RelationshipMeta {
  value: RelationshipType;
  label: string;
  short: string;
}

export const RELATIONSHIPS: readonly RelationshipMeta[] = [
  { value: 'direct_report', label: 'Liderado direto', short: 'Liderados' },
  { value: 'peer',          label: 'Par',             short: 'Pares' },
  { value: 'manager',       label: 'Chefe',           short: 'Chefes' },
  { value: 'skip_level',    label: 'Skip-level',      short: 'Skip' },
  { value: 'stakeholder',   label: 'Stakeholder',     short: 'Stakeholders' },
  { value: 'other',         label: 'Outro',           short: 'Outros' },
] as const;

export const RELATIONSHIP_LABEL: Record<RelationshipType, string> = Object.fromEntries(
  RELATIONSHIPS.map((r) => [r.value, r.label]),
) as Record<RelationshipType, string>;

export interface Person {
  id: string;
  name: string;
  email: string;
  relationship: RelationshipType;
  role?: string | null;
  slack_id?: string | null;
  jira_account_id?: string | null;
  github_handle?: string | null;
  start_date?: string | null;
  notes?: string | null;
  created_at: string;
  updated_at: string;
}

export interface PersonDraft {
  id: string;
  name: string;
  email: string;
  relationship: RelationshipType;
  role?: string | null;
  slack_id?: string | null;
  jira_account_id?: string | null;
  github_handle?: string | null;
  start_date?: string | null;
  notes?: string | null;
  source: string;
  proposed_at: string;
}

export type PersonInput = Omit<Person, 'id' | 'created_at' | 'updated_at'>;
export type PersonUpdate = Partial<PersonInput>;
